pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java

```diff
@@ -0,0 +1,121 @@
+import org.axonframework.spring.stereotype.Saga;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+import java.util.UUID;
+
+@Saga
+@Revision("0")
+@Log4j2
+@NoArgsConstructor
+public class OrganizationRegistrationSaga {
+    private static final String ASSOCIATION_PROPERTY = "registrationConfirmationCode";
+
+    private transient CommandGateway commandGateway;
+    private transient UsersReadService usersReadService;
+
+    private UUID organizationId;
+    private UUID registeringUserId;
+    private String registeringUserEmail;
+    private String registeringUserEncodedPassword;
+    private String registeringUserDisplayName;
+    private String organizationName;
+    private String street;
+    private String city;
+    private String state;
+    private String country;
+    private String postalCode;
+    private String websiteUrl;
+    private String phoneNumber;
+
+    // Not implemented: deadline management for handling confirmation timeouts
+
+    @Autowired
+    @Qualifier("hazelcastCommandGateway")
+    public void setCommandGateway(CommandGateway commandGateway) {
+        this.commandGateway = commandGateway;
+    }
+
+    @Autowired
+    public void setUsersReadService(UsersReadService usersReadService) {
+        this.usersReadService = usersReadService;
+    }
+
+    @StartSaga
+    @SagaEventHandler(associationProperty = ASSOCIATION_PROPERTY)
+    public void on(OrganizationRegistrationReceivedEvent event) {
+        organizationId = event.getOrganizationId();
+        registeringUserId = event.getRegisteringUserId();
+        registeringUserEncodedPassword = event.getRegisteringUserEncodedPassword();
+        registeringUserDisplayName = event.getContactName();
+        registeringUserEmail = event.getRegisteringContactEmail();
+        organizationName = event.getOrganizationName();
+        street = event.getStreet();
+        city = event.getCity();
+        state = event.getState();
+        country = event.getCountry();
+        postalCode = event.getPostalCode();
+        websiteUrl = event.getWebsiteUrl();
+        phoneNumber = event.getContactPhoneNumber();
+
+        // Add an email to your outbound queue.... maybe record the message ID in the command
+
+        commandGateway.send(new RecordSentOrganizationRegistrationEmailConfirmationCommand(event.getRegistrationConfirmationCode(),
+                event.getOrganizationId(), event.getRegisteringContactEmail(), event.getOrganizationName()));
+    }
+
+    @SagaEventHandler(associationProperty = ASSOCIATION_PROPERTY)
+    public void on(OrganizationRegistrationConfirmedEvent event) {
+        // The business may require that each email address be only available for registration once. In this case,
+        // the target aggregate identifier may be the user email address (possibly with an additional prefix).
+        if (usersReadService.hasUserWithEmail(registeringUserEmail)) {
```

Comment:
As we discussed before, this can possibly have a racing condition. Without a choking point, e.g. synchronisation, it is unlikely we can achieve total thread safety. Obviously, the downside of introducing synchronisation is performance. 

The tricky bit of this racing condition is that it could be totally random. There is no guarantee that during a replay the order would be the same, i.e. if Client_A succeeds and Client_B fails during normal application processing, when the system replays, it is possible that Client_A fails and Client_B succeeds!? This could be wreaking havoc.

Reply:
If I understand the question correctly, your concern is that different commands will be dispatched during a replay? This won't happen as sagas do not handle events during replay.

Reply:
Ah OK I wasn't aware that Saga event handlers are *not* part of replay. It makes sense now.